In [1]:
print('hi')

hi


In [2]:
import os
import scanpy as sc
import pandas as pd
import cellxgene_census

In [32]:
#input_dir = "/projects/bioinformatics/DB/scRNAseq_parkinson"
# the full path to the .h5ad file
#h5ad_path = os.path.join(input_dir, "dataset.h5ad")

In [3]:
 metadata = pd.read_csv('/projects/bioinformatics/DB/scRNAseq_parkinson/metadata.csv')

In [4]:
metadata.columns

Index(['barcodekey', 'n_genes', 'n_counts', 'Brain_bank', 'RIN',
       'path_braak_lb', 'derived_class2', 'PMI', 'tissue_ontology_term_id',
       'tissue_type', 'assay_ontology_term_id', 'disease_ontology_term_id',
       'cell_type_ontology_term_id',
       'self_reported_ethnicity_ontology_term_id',
       'development_stage_ontology_term_id', 'sex_ontology_term_id',
       'donor_id', 'suspension_type', 'is_primary_data', 'cell_type', 'assay',
       'disease', 'sex', 'tissue', 'self_reported_ethnicity',
       'development_stage', 'observation_joinid'],
      dtype='object')

In [5]:
metadata['observation_joinid']


0          MpT)aB)#*U
1          qFT~wG^BCg
2          fE`VpY%8fb
3          g1Lr8^y)O(
4          MR9tUCifao
              ...    
2096150    25FK4DT!Rl
2096151    PdMVWS9R^v
2096152    Q&?JBHEFAz
2096153    @vUY6<XF#a
2096154    *Ck9$@#b{s
Name: observation_joinid, Length: 2096155, dtype: object

In [6]:
obs_join_ids = list(metadata['observation_joinid'])
#obs_value_filter = 'observation_joinid in @obs_join_id'
obs_value_filter = f'observation_joinid in {obs_join_ids}'

In [101]:
obs_value_filter;

In [7]:
census = cellxgene_census.open_soma(census_version="2025-01-30")

# Meta Data 

In [8]:
cell_metadata = cellxgene_census.get_obs(
    census=census,
    organism="homo_sapiens",
    value_filter=obs_value_filter,
    column_names=["observation_joinid","cell_type"]
)

In [9]:
cell_metadata;

In [10]:
# Some PBMC data from this data 
# 1. https://datasets.cellxgene.cziscience.com/88ad4023-93b2-4b70-a8a5-b0cf6c438cac.h5ad
#https://cellxgene.cziscience.com/collections/d5d0df8f-4eee-49d8-a221-a288f50a1590 <--- this we need to use

import cellxgene_census


emb_names = ["scvi", "geneformer"]


census_version = "2025-01-30"


#dataset_ids = ["d5d0df8f-4eee-49d8-a221-a288f50a1590"]


with cellxgene_census.open_soma(census_version=census_version) as census:
    adata = cellxgene_census.get_anndata(
        census,
        organism="homo_sapiens",
        measurement_name="RNA",
        #obs_value_filter='dataset_id in ["d5d0df8f-4eee-49d8-a221-a288f50a1590"]',
        obs_embeddings=emb_names,
        obs_value_filter=obs_value_filter,
        obs_column_names = ["observation_joinid"]
    )


In [11]:
adata.obsm

AxisArrays with keys: scvi, geneformer

In [12]:
df_scvi = pd.DataFrame(
    adata.obsm["scvi"], 
    index=adata.obs_names
)
df_scvi.to_csv("scvi_embeddings.csv")

In [13]:
df_scvi

0         1         2         3         4             5   \
0       -0.004639  0.004547  0.215820  0.000767 -0.003586 -5.836487e-04   
1        0.000637  0.002472 -0.367188  0.001457 -0.003937 -2.151489e-03   
2       -0.002014  0.003387  0.188477 -0.001556 -0.001930 -1.823425e-03   
3       -0.001549  0.002762 -0.699219 -0.001869  0.000957 -1.289368e-03   
4        0.001930  0.000136 -0.925781 -0.006287  0.000839  3.738403e-03   
...           ...       ...       ...       ...       ...           ...   
2096150 -0.000866  0.003067  0.435547 -0.003006 -0.001381 -3.021240e-03   
2096151  0.006409  0.002167  0.058838  0.002426  0.000874 -5.699694e-07   
2096152  0.000675  0.001602  0.482422  0.001724 -0.005432 -3.738403e-03   
2096153  0.004181 -0.006500  0.195312  0.004181  0.002335  4.547119e-03   
2096154 -0.004242  0.005249  0.271484 -0.000355 -0.006866 -3.097534e-03   

               6         7         8         9   ...        40        41  \
0       -0.175781  0.333984 -0.001541 -0.003586  ...  0.001129 -0.351562   
1       -1.039062 -0.124023 -0.000167 -0.001587  ... -0.001511 -0.037109   
2       -0.644531 -0.241211 -0.000683  0.002594  ... -0.000984 -0.322266   
3       -0.484375 -0.076172  0.000041  0.001183  ... -0.003021  0.287109   
4        0.664062  0.914062 -0.001656  0.001312  ... -0.004089  0.398438   
...           ...       ...       ...       ...  ...       ...       ...   
2096150 -0.341797  0.159180 -0.000637 -0.000043  ... -0.001122 -0.058350   
2096151 -0.812500 -0.550781  0.004791  0.004364  ...  0.002930 -0.427734   
2096152 -0.765625  2.140625  0.000587 -0.011353  ... -0.002365  1.054688   
2096153  1.750000  0.136719 -0.003387 -0.000032  ... -0.000209 -0.115723   
2096154 -0.486328  1.546875  0.002029 -0.005005  ... -0.001244  0.185547   

               42        43        44        45        46        47        48  \
0       -0.003418 -0.000984  0.080566 -0.002121  0.464844  0.002518 -0.000188   
1        0.004395 -0.001526  0.031494  0.003479  0.921875  0.004028  0.005829   
2       -0.004150 -0.002365  0.157227  0.001373  0.597656  0.003510  0.001068   
3       -0.000553 -0.000504  0.186523  0.002701  0.753906  0.003204  0.000412   
4        0.005280  0.002396 -0.235352  0.003754  1.117188  0.000706  0.000736   
...           ...       ...       ...       ...       ...       ...       ...   
2096150 -0.001122 -0.002472 -0.144531 -0.002213  0.593750  0.001999 -0.001297   
2096151  0.000113  0.004211  0.562500  0.006561 -2.015625  0.000889  0.001968   
2096152 -0.000340 -0.003113  0.235352 -0.008301  0.644531 -0.001823  0.001587   
2096153  0.004028  0.001289  1.007812 -0.007019 -0.022339 -0.001122 -0.003845   
2096154 -0.004364 -0.002960  0.084473 -0.002884  0.617188  0.003418  0.004059   

               49  
0       -0.004028  
1        0.001801  
2        0.001114  
3       -0.003784  
4       -0.002579  
...           ...  
2096150 -0.002502  
2096151  0.005219  
2096152 -0.004272  
2096153 -0.002792  
2096154 -0.001648  

[2096155 rows x 50 columns]

In [87]:
print(adata.obsm.keys())

KeysView(AxisArrays with keys: scvi, geneformer)


In [88]:
print(adata.shape)  
print(adata.obs.head())

(0, 61888)
Empty DataFrame
Columns: [observation_joinid, dataset_id]
Index: []


In [18]:
df_scvi.shape

(2096155, 50)

In [14]:
df_geneformer = pd.DataFrame(
    adata.obsm["geneformer"], 
    index=adata.obs_names
)
df_geneformer.to_csv("geneformer_embeddings.csv")


In [15]:
df_geneformer.head()

0         1         2         3         4         5         6    \
0  0.121582 -0.871094 -0.396484  0.726562 -0.746094  0.267578  0.170898   
1  0.550781 -1.140625  0.012146  0.437500 -0.968750 -0.048584  0.150391   
2 -0.093750 -1.851562  0.316406 -0.691406 -0.871094 -0.035156  0.202148   
3  0.039062 -2.015625  0.009949  0.519531 -1.148438 -0.000176  0.625000   
4  0.527344 -1.343750  0.554688  1.023438 -0.221680 -0.519531  0.104492   

        7         8         9    ...       502       503       504       505  \
0  1.382812  0.964844  0.511719  ... -0.808594 -1.250000  0.593750 -0.490234   
1  0.687500  0.291016  1.109375  ... -0.271484 -0.570312  0.091309 -0.470703   
2 -0.015076  0.410156  0.515625  ...  0.792969 -0.710938  0.337891  0.593750   
3  0.291016  0.636719  1.054688  ... -0.119141 -0.570312 -0.084473  0.287109   
4  0.738281 -0.328125  0.527344  ... -0.134766 -0.400391  0.056641 -0.667969   

        506       507       508       509       510       511  
0  0.257812  0.968750 -0.843750 -0.226562  0.099609 -0.328125  
1 -0.159180  1.109375 -0.406250  0.184570 -0.257812 -0.369141  
2  0.032227  0.141602  0.343750  0.867188 -0.671875 -0.625000  
3  0.138672  0.447266 -0.105469  1.164062 -0.337891 -0.875000  
4  0.482422  1.132812 -0.734375 -0.269531  0.310547 -0.417969  

[5 rows x 512 columns]

In [17]:
df_geneformer.shape

(2096155, 512)

In [4]:
# Get the size in bytes
size_bytes = os.path.getsize(h5ad_path)

# Convert to a more readable unit (e.g., MB or GB)

size_gb = size_bytes / (1024 * 1024 * 1024)

print(f"Size in gigabytes: {size_gb:.2f} GB")

Size in gigabytes: 28.45 GB


In [6]:
import zarr

### Size issue 
That file size of 28.45 GB for a single-cell RNA-seq .h5ad dataset is very large and will likely pose significant computational challenges for a standard analysis,
especially when using a Transformer-based model like GenFormer.

Memory (RAM): A 28 GB .h5ad file, which typically contains a sparse matrix of UMI counts, will require a substantial amount of RAM when loaded and processed. When the data is converted to a dense matrix or intermediate processing steps are performed (like calculating nearest neighbors, running PCA, etc.), the memory footprint can easily exceed 100-200 GB. If you don't have access to a high-memory computing environment (e.g., a powerful server or cloud instance), the analysis will crash.

Processing Time: Even with sufficient RAM, a dataset of this size (likely involving millions of cells) will result in long processing times for any single-cell workflow, particularly for computationally intensive steps like training a large neural network model such as GenFormer.